Fix bug where checking for broken links always failed due to

coreysals opened the pull request:
changes in "next"/"previous" link generation mistakenly adding escaped
back slash rather than, correctly, forward slash.

This bug was previously introduced when code was modified to allow
unique .md file names instead of enforcing "section.md" file name.

Issue:
https://alliancels.myjetbrains.com/youtrack/issue/DT-51

External project test branch on maven repo for doc tools changes:
https://github.com/alliancels/maven-repo/tree/DT51_ExternalTestBranch

Titan branch with fixed links and gradle build file set to use above external project test branch:
https://github.com/alliancels/Titan/tree/TIT770_FixMissedDocToolsBrokenLinks
(I'll create a separate pull request to review these documentation changes)

coreysals committed:
Fix bug where checking for broken links always failed due to

changes in "next"/"previous" link generation mistakenly adding escaped
back slash rather than, correctly, forward slash.

This bug was previously introduced when code was modified to allow
unique .md file names instead of enforcing "section.md" file name.

benhammen reviewed (changes requested):
The checker now works, but the problem reported in the YouTrack issue still basically exists: "Documentation tool does not check for broken links".  It technically *can*, but it won't be done by anyone, due to documentation problems.

The instructions on Sharepoint and in the Titan/Docs/README exclude the checker. The Sharepoint instructions are also ridiculously excessive for something that should only take one command (or possibly two, if a clean is really needed), plus they incorrectly duplicate what's documented elsewhere, so I just deleted them.

```
User@machine $ gradlew clean  
User@machine $ gradlew --refresh-dependencies  
User@machine $ gradlew Docs:aA --rerun-tasks  
User@machine $ gradlew Docs:combineAll 
```

The doc tool does have a build task that combines assembling and checking, but it doesn't currently chain in the "combine" task.  That can very easily be added here: https://github.com/alliancels/documentation-gradle-plugins/blob/master/src/main/groovy/com/alliancels/documentation/DocumentationPlugin.groovy#L94. 

The doc tool project README can then be updated to mention the build task also includes the combine task: https://github.com/alliancels/documentation-gradle-plugins#build-documentation.

The Titan Doc project README should later be updated with better/simpler instructions, when the Titan repo is updated to use the new tool version.  Again, clean and build should be all that's mentioned.

coreysals committed:
Add combineAll task to buildAll task.

Update tasks so that following a clean tasks that were previously
skipped due to "up-to-date" cached results are now ran.

coreysals committed:
Update README to describe combine task being included in build task.

coreysals commented:
I had initially left the combine task out thinking that maybe people wouldn't always care to generate the combined docs and would rather save the build time. That being said I don't think either of those initial worries are warranted. 

I made further modifications to DocumentationPlugin.groovy in an attempt to better ensure that following a 'clean', 'buildAll' would run all appropriate tasks instead of skipping some due to output cache existing from previous task completion(s). I'm not confident this is the best solution to this problem however I think something is needed beyond expecting the user to invoke --rerun-tasks.

benhammen reviewed (changes requested):
Skipping tasks when an output folder or links already exist will require always doing a clean before a build, or risk having the output be out-of-date, which is also annoying, and could even lead to false broken link errors.

If you just set all tasks to `outputs.upToDateWhen {false}` , then we won't need "clean" or "--rerun-tasks".  I think that may be the best solution.

The incremental build never worked, due to some buggy/flaking behavior with the Gradle API.  It's possible that could be fixed now (in the newer Gradle version), and can be setup properly, but I'm not sure if it's even worth the effort.  If that's desired, a new Youtrack issue should probably be created for that.

coreysals committed:
Change all outputs.upToDateWhen to false.

Done as simple solution to avoid issues involving cached outputs
following a clean or "rebuild".

coreysals committed:
Remove console prints for combining docs and linking glossary.

These were initially left in place in an attempt to help inform of the
combine/glossary/link features added to the doc tools. Removed for
consistency with other tasks.

coreysals commented:
I'll squash the "Change all outputs..." with the "Add combineAll..." commits at merge. 

I added another commit to remove the console feedback of the combine task because I don't think the feedback was useful and this is more consistent with the other tasks. This is more cleanup so maybe I shouldn't lump it in with this branch last minute. Let me know what you think.

benhammen commented:
Everything looks good to me.

Note that I did not actually test out this most recent version, so hopefully you at least did a basic sanity check on the example project.

coreysals commented:
> Note that I did not actually test out this most recent version, so hopefully you at least did a basic sanity check on the example project.

I did and I'll push to the external test branch and run on Titan before merging, versioning, and ultimately update Titan and Titan README.  

Thanks!

coreysals commented:
Merged. Thanks!
